From 5acb8b5f5968dca7b7d34f1e7c8c634a6bfff21d Mon Sep 17 00:00:00 2001
From: Sebastian Schaper <openwrt@sebastianschaper.net>
Date: Sun, 28 Mar 2021 00:32:46 +0100
Subject: [PATCH 2/3] ath79: add support for ZyXEL NWA1123-NI

Specifications:
 * AR9342, 16 MiB Flash, 64 MiB RAM, 802.11n 2T2R, 2.4 GHz
 * AR9382 PCIe card, 802.11n 2T2R, 5 GHz
 * 1x Gigabit Ethernet (AR8035), 802.3af PoE

Installation:
* OEM Web UI is at 192.168.1.2
  login as `admin` with password `1234`
* Flash factory-AAEO.bin

The string `AAEO` needs to be present within the file name of the uploaded
image to be accepted by the OEM Web-based updater, the factory image is
named accordingly to save the user from the hassle of manual renaming.

TFTP Recovery:
* Open the case, connect to TTL UART port (this is the official method
  described by Zyxel, the reset button is useless during power-on)
* Extract factory image (.tar.bz2), serve `vmlinux_mi124_f1e.lzma.uImage`
  and `mi124_f1e-jffs2` via tftp at 192.168.1.10
* Interrupt uboot countdown, execute commands
  `run lk`
  `run lf`
  to flash the kernel / filesystem accordingly

MAC addresses as verified by OEM firmware:
use   address   source
LAN   *:fb      mib0 0x30 ('eth0mac'), art 0x1002 (label)
2g    *:fc      mib0 0x4b ('wifi0mac')
5g    *:fd      mib0 0x66 ('wifi1mac')

Signed-off-by: Sebastian Schaper <openwrt@sebastianschaper.net>
---
 target/linux/ath79/dts/ar9342_zyxel_nwa1123-ni.dts   | 12 ++++++++++++
 .../ath79/generic/base-files/etc/board.d/02_network  |  3 ++-
 .../etc/hotplug.d/ieee80211/10_fix_wifi_mac          |  6 ++++++
 target/linux/ath79/image/generic.mk                  |  8 ++++++++
 4 files changed, 28 insertions(+), 1 deletion(-)
 create mode 100644 target/linux/ath79/dts/ar9342_zyxel_nwa1123-ni.dts

diff --git a/target/linux/ath79/dts/ar9342_zyxel_nwa1123-ni.dts b/target/linux/ath79/dts/ar9342_zyxel_nwa1123-ni.dts
new file mode 100644
index 0000000000..1f5ae76aa0
--- /dev/null
+++ b/target/linux/ath79/dts/ar9342_zyxel_nwa1123-ni.dts
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "ar9342_zyxel_nwa112x.dtsi"
+
+/ {
+	compatible = "zyxel,nwa1123-ni", "qca,ar9342";
+	model = "Zyxel NWA1123-NI";
+};
+
+&pcie {
+	status = "okay";
+};
diff --git a/target/linux/ath79/generic/base-files/etc/board.d/02_network b/target/linux/ath79/generic/base-files/etc/board.d/02_network
index 0b919ee9ad..bacae74f7a 100644
--- a/target/linux/ath79/generic/base-files/etc/board.d/02_network
+++ b/target/linux/ath79/generic/base-files/etc/board.d/02_network
@@ -93,7 +93,8 @@ ath79_setup_interfaces()
 	ubnt,unifi|\
 	wd,mynet-wifi-rangeextender|\
 	winchannel,wb2000|\
-	zyxel,nwa1121-ni)
+	zyxel,nwa1121-ni|\
+	zyxel,nwa1123-ni)
 		ucidef_set_interface_lan "eth0"
 		;;
 	airtight,c-75)
diff --git a/target/linux/ath79/generic/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac b/target/linux/ath79/generic/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac
index 0d8a15a780..ac1285b241 100644
--- a/target/linux/ath79/generic/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac
+++ b/target/linux/ath79/generic/base-files/etc/hotplug.d/ieee80211/10_fix_wifi_mac
@@ -68,4 +68,10 @@ case "$board" in
 	zyxel,nwa1121-ni)
 		mtd_get_mac_text mib0 0x4b > /sys${DEVPATH}/macaddress
 		;;
+	zyxel,nwa1123-ni)
+		[ "$PHYNBR" -eq 0 ] && \
+			mtd_get_mac_text mib0 0x4b > /sys${DEVPATH}/macaddress
+		[ "$PHYNBR" -eq 1 ] && \
+			mtd_get_mac_text mib0 0x66 > /sys${DEVPATH}/macaddress
+		;;
 esac
diff --git a/target/linux/ath79/image/generic.mk b/target/linux/ath79/image/generic.mk
index 71b2e5d945..730fb33f19 100644
--- a/target/linux/ath79/image/generic.mk
+++ b/target/linux/ath79/image/generic.mk
@@ -2261,6 +2261,14 @@ define Device/zyxel_nwa1121-ni
 endef
 TARGET_DEVICES += zyxel_nwa1121-ni
 
+define Device/zyxel_nwa1123-ni
+  $(Device/zyxel_nwa112x)
+  DEVICE_MODEL := NWA1123
+  DEVICE_VARIANT := NI
+  ZYXEL_MODEL_STRING := AAEO
+endef
+TARGET_DEVICES += zyxel_nwa1123-ni
+
 define Device/zyxel_nbg6616
   SOC := qca9557
   DEVICE_VENDOR := ZyXEL
-- 
2.29.2.windows.2

